File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py


def solution(m, n, board):
    board = list(map(list, list(zip(*board))))
    return remove(board)

def remove(board):  
    answer = 0
    delete = set()
    n = len(board)
    m = len(board[0])

    for j in range(n - 1):
        for i in range(m - 1):
            if board[j][i]:
                if is_square(board, j, i):
                    delete.add((j, i))
                    delete.add((j, i + 1))
                    delete.add((j + 1, i))
                    delete.add((j + 1, i + 1))
                    
    
    if not delete:
        return answer
    
    answer += len(delete)
    for y, x in delete:
        board[y][x] = ""
    
    for y in range(len(board)):
        b = list("".join(board[y]))
        board[y] = [""] * (m - len(b)) + b
    
    return answer + remove(board)
        
    
        
        
    
def is_square(board, y, x):
    dy = [0, 1, 1]
    dx = [1, 1, 0]
    
    target = board[y][x]
    if not target:
        return False
    
    for i in range(3):
        if target != board[y+dy[i]][x+dx[i]]:
            return False
        
    return True
            
    
    
                
                
                

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int m, int n, String[] board) {
        char[][] b = new char[board[0].length()][board.length];
        
        for (int j = 0; j < board.length; j++) {
            for (int i = 0; i < board[j].length(); i++) {
                b[i][j] = board[j].charAt(i);
            }
        }
        
        return remove(b);
    }
    
    public int remove(char[][] board) {
        int answer = 0;
        int n = board.length;
        int m = board[0].length;
        boolean[][] toRemove = new boolean[n][m];
        
        for (int j = 0; j < n - 1; j++) {
            for (int i = 0; i < m - 1; i++) {
                if (board[j][i] != '-' && isSquare(board, j, i)) {
                    toRemove[j][i] = true;
                    toRemove[j][i + 1] = true;
                    toRemove[j + 1][i] = true;
                    toRemove[j + 1][i + 1] = true;
                }
            }
        }
        
        for (int y = 0; y < n; y++) {
            for (int x = 0; x < m; x++) {
                if (toRemove[y][x]) {
                    board[y][x] = '-';
                    answer++;
                }
            }
        }
        
        for (int y = 0; y < board.length; y++) {
            char[] b = board[y];
            StringBuilder sb = new StringBuilder();
            
            for (char c : b) {
                sb.append(c);    
            }
            
            String s = Arrays.stream(sb.toString()
                          .split("-"))
                .collect(Collectors.joining());
            
            s = "-".repeat(m - s.length()) + s;
            
            for (int i = 0; i < m; i++) {
                board[y][i] = s.charAt(i);
            }                        
        }
        
        return answer == 0 ? 0 : answer + remove(board);
        
    }
    
    public boolean isSquare(char[][] board, int y, int x) {
        int[] dy = new int[]{0,1,1};
        int[] dx = new int[]{1,1,0};
        
        char target = board[y][x];
        
        if (target == '-') {
            return false;
        }
        
        for (int i = 0; i < 3; i++) {
            if (target != board[y + dy[i]][x + dx[i]]) {
                return false;
            }
        }
        return true;
    }
}